api/pkg/webhooks/client.go

```diff
@@ -0,0 +1,180 @@
+// package webhooks provides a webhook manager that can be used to invoke webhooks for different events.
+
+/*
+Usage:
+1. In the caller package (eg, mlp, merlin), define the list of events that requires webhooks. For example:
+
+```go
+const (
+	ProjectCreatedEvent wh.EventType = "OnProjectCreated"
+	ProjectUpdatedEvent wh.EventType = "OnProjectUpdated"
+)
+
+var EventList = []wh.EventType{
+	ProjectCreatedEvent,
+	ProjectUpdatedEvent,
+}
+```
+
+2. Define the event to webhook configuration. Optionally, the configuration can be provided in a yaml file
+and parsed via the `Config` struct.
+In the config file, define the event to webhook mapping for those events as required.
+For example, if projects need extra labels from an external source,
+we define the webhook config for the `OnProjectCreated` event
+
+```go
+webhooks:
+  enabled: true
+  config:
+    OnProjectCreated:
+      - url: http://localhost:8081/project_created
+        method: POST
+        onError: abort
+```
+
+3. Call InitializeWebhooks() to get a WebhookManager instance.
+   This method will initialize the webhook clients for each event type based on the mapping provided
+
+```go
+projectsWebhookManager, err := webhooks.InitializeWebhooks(cfg.Webhooks, service.EventList)
+```
+
+4. Call `InvokeWebhooks()` method in the caller code based on the event
+*/
+
+package webhooks
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/avast/retry-go/v4"
+	"github.com/caraml-dev/mlp/api/log"
+)
+
+type EventType string
+type ServiceType string
+
+const (
+	onErrorIgnore = "ignore"
+	onErrorAbort  = "abort"
+)
+
+type WebhookClient interface {
+	Invoke(context.Context, []byte) ([]byte, error)
+	InvokeAsync(context.Context, []byte) error
+	IsAsync() bool
+	AbortOnFail() bool
+	IsFinalResponse() bool
+	GetUseDataFrom() string
+	GetName() string
+}
+
+type simpleWebhookClient struct {
```

Comment:
actually structurally it feels bit off to me after looking deeper. Base on the interface, all `WebhookClient` has both `Invoke` and `InvokeAsync` but in reality its either or and you will use one of them only. 

```
type WebhookClient interface {
	Invoke(context.Context, []byte) ([]byte, error)
	AbortOnFail() bool
	IsFinalResponse() bool
	GetUseDataFrom() string
	GetName() string
}

type WebhookClientBase struct {
...
}

type SyncWebhookClient struct {
	WebhookClientBase
}

func (s *SyncWebhookClient) Invoke(ctx context.Context, data []byte) ([]byte, error) {
	// Synchronous invocation logic here
	fmt.Println("Synchronous Invoke")
	return data, nil
}

type AsyncWebhookClient struct {
	WebhookClientBase
}

func (a *AsyncWebhookClient) Invoke(ctx context.Context, data []byte) ([]byte, error) {
	// Asynchronous invocation logic here
	fmt.Println("Asynchronous Invoke")
```

These said, i think from the client perspective, it doesn't need to have any form of knowledge of sync/async, we can simply merge the Invokes into and keep the `isAsync` on the config but not this class. That way from the way we construct the webhook flow, we still know how to put them into the async/sync, without this client being aware.  

The extra methods and flag give untrue hints about how the client works (should i use channel for invokeasync? can i use a client in both async and sync fashion?)

Reply:
Actually, the user of this library will not need to call the `WebhookClient` directly, there should interact primarily with the `WebhookManager` interface and call the `InvokeWebhooks()` method. The `WebhookClient` interface is exposed for the user to mock the return value from the `Invoke()` method for testing.

> all `WebhookClient` has both `Invoke` and `InvokeAsync`

because it is up to the user to decide whether a particular call should be synchronous or async. But i see your point, the client can be instantiated based on the configuration, depending on whether the async flag is set or not.

Reply:
yep the client dont even need to be concern with it being async or not, the manager can orchestrate it